### PR TITLE
Fixed severals bugs

### PR DESCRIPTION
**Pull request by Kyew**
Fixed zg_context_new_graph_from_string(ZGContext *, const char *) by fixing the constructor PdFileParser::PdFileParser(std::string) which was not initializing the 'pos' member, leading to the string being partially parsed or not parsed at all.

Fixed PdGraph::resolveFullPath(const char *) returning NULL when the graph as no parent whereas it should return a std::string

Fixed zg_graph_unattach(PdGraph *graph) directly calling PdGraph::attachToContext(bool) instead of calling PdContext::unattachGraph(PdGraph *). Also fixed PdContext::unattachGraph(PdGraph *) not removing the graph to its graph list.

Implemented MessageSoundfiler method, now it works but only support the read command

Fixed DspObject::getConnectionType(int) always returning DSP where it could be a MESSAGE outlet

### CODE REVIEW

**Comment by Kyew** [comment]
I fixed another bug, I updated the pull request comment to describe it.

**Comment by Kyew** [comment]
Fixed another bug, cf. pull request first comment

File: src/DspObject.cpp [hunk 1]
```diff
@@ -71,6 +71,8 @@ DspObject::~DspObject() {
 #pragma mark -
 
 ConnectionType DspObject::getConnectionType(int outletIndex) {
+  if (outletIndex >= getNumDspOutlets())
```
Comment (mhroth) [inline src/DspObject.cpp]: Why is this necessary? Why would `outletIndex` ever be greater than `getNumDspOutlets()`?

File: src/MessageSoundfiler.cpp [hunk 2]
```diff
@@ -41,96 +42,108 @@ const char *MessageSoundfiler::getObjectLabel() {
...
+    sendMessage(0, outgoingMessage);
   }
-*/
-}
+  else if (message->isSymbol(0, "write"))
+  {
+    //Not implemented yet
+    graph->printErr("[soundfiler]: The 'write' command is not supported yet.");
+  }
+}
```
Comment (mhroth) [inline src/MessageSoundfiler.cpp]: Add an empty line at the end of the file.

File: src/MessageSoundfiler.cpp [hunk 3]
```diff
@@ -2,7 +2,7 @@
  *  Copyright 2009,2010 Reality Jockey, Ltd.
  *                 info@rjdj.me
  *                 http://rjdj.me/
- *
```
Comment (mhroth) [inline src/MessageSoundfiler.cpp:5]: I'll assume that all of this works...!

File: src/DspObject.cpp (hunk 1 above)
Comment (Kyew) [inline src/DspObject.cpp]: Some objects in pure-data have dsp inlets, but only message outlets. For example, I implemented, as an external object, the threshold~ pure-data object. It has one dsp inlet (so I needs to be a DspObject), two messages inlets, two messages outlets but no dsp outlet.

Comment (mhroth) [inline src/DspObject.cpp]: Ahh, I see. In that case when you write your `DspThreshold` object, declare it as `DspThreshold(2, 1, 2, 0)` (if this is a object with two inlets, but the left one can take signal and control) or `DspThreshold(3, 1, 2, 0)` (if the left inlet should be signal only) and override the `getConnectionType` function to return `MESSAGE`. Don't add non-general behaviour `DspObject::getConnectionType()`.

File: src/MessageSoundfiler.cpp (hunk 2 above)
Comment (Kyew) [inline src/MessageSoundfiler.cpp]: It's done.

File: src/DspObject.cpp (hunk 1 above)
Comment (Kyew) [inline src/DspObject.cpp]: Ok, I didn't think about that solution. I reverted this commit.

File: src/PdContext.cpp [hunk 4]
```diff
@@ -206,7 +206,8 @@ void PdContext::attachGraph(PdGraph *graph) {
 
 void PdContext::unattachGraph(PdGraph *graph) {
   lock();
-  //graphList.erase(graph); // TODO(mhroth): remove the graph from the graphList
+  graphList.erase(std::remove(graphList.begin(), graphList.end(), graph),
+                  graphList.end());
```
Comment (mhroth) [inline src/PdContext.cpp]: One last (very small) issue. Line 210 should be indented only 4 spaced, not lined up with the parenthesis (Xcode does this automatically. I hate it). That will make the formatting consistent with the rest of the code. After this, I'm happy to merge everything in.

**Comment by Kyew** [comment]
I fixed the indentation
